### PR TITLE
feat(text-input): add readonly variant

### PR DESCRIPTION
**Pull request by janhassel**
Ref #2177, #3018, #3054

Adds a readonly variant to the TextInput component.

Intentional differences from the original design spec:
- No tooltip on the `EditOff16` icon
  - Looking at the planned changes of the tooltip component in v11, adding full support for that at this time seems inefficient. It would require the addition of multiple props that aren't really in the scope of the text input but rather of the tooltip (`tooltipAlignment`, `tooltipPosition`) and might introduce accessiblity issues that are to be addressed with the tooltip changes in v11.
  - In the meantime, teams can still inform their users about the reason why a field is read-only with a couple of options:
    - Use the helper text as explaination
    - Include a tooltip in the form label (quick poc with disabled field: https://jfckr.csb.app/)
- No tooltip on the entire input to show overflow text
  - Same as above
  - Users have the ability to "scroll" inside the field with both their mouse or their keyboard (arrow keys) to reveal overflowing text (native browser behaviour)

#### Changelog

**New**

- `props.readonly` on `TextInput`

#### Testing / Reviewing

- Ensure `readonly` attribute is set to `<input>` accordingly
- Ensure no visual regressions appear with any combination of `invalid`, `warn`, `readonly` on both the `TextInput` and the `PasswordInput` component
- Ensure that no error states are shown when a field is marked as readonly

Side note: this PR also adds a helper `useNormalizedInputProps` which aims to reduce some of the many conditional renderings that were introduced over time (e.g. `{!invalid && !warn && !isFluid && !inline && helper}`) by outsourcing them and providing a single entry for the icons and validation messages.
The intent is to add readonly support to other inputs as well (like `NumberInput`). This utility should help there as well and streamline the inputs. Let me know what you think!

### CODE REVIEW

File: packages/react/src/components/TextInput/TextInput.js [hunk 1]
```diff
@@ -241,6 +238,11 @@ TextInput.propTypes = {
    */
   placeholder: PropTypes.string,
 
+  /**
+   * Whether the input should be read-only
+   */
+  readonly: PropTypes.string,
```
Comment (sstrubberg) [inline packages/react/src/components/TextInput/TextInput.js]:
```suggestion
  readonly: PropTypes.bool,
```

image

☝️ When the PropType is set to string, I get the following in my console. When I turn it into a bool it goes away

**Review by sstrubberg — CHANGES_REQUESTED** [review]
Read-only appears to take precedence over warn and invalid. Was that the intended outcome?

Also, the stuff below.

File: packages/react/src/components/TextInput/TextInput.js (hunk 1 above)
Comment (janhassel) [reply packages/react/src/components/TextInput/TextInput.js]: Thanks for catching that!

**Comment by janhassel** [comment]
@sstrubberg Yes, when `readonly` is passed, both `invalid` and `warn` should be ignored. I based this on @shixiedesign's [comment from two years ago](https://github.com/carbon-design-system/carbon/pull/3054#pullrequestreview-251780808):
> Is it possible to disable "error state" on read-only? This situ shouldn't arise. It doesn't give user a way forward since field cannot be edited.

**Review by sstrubberg — APPROVED** [review]

File: packages/react/src/internal/useNormalizedInputProps.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import {
+  WarningFilled16,
+  WarningAltFilled16,
+  EditOff16,
+} from '@carbon/icons-react';
+import { settings } from 'carbon-components';
+
+const { prefix } = settings;
+
+export function useNormalizedInputProps({
```
Comment (tay1orjones) [inline packages/react/src/internal/useNormalizedInputProps.js:53]: I think this is a good point of abstraction. Could you add a jsdoc-type comment for it so we have some context on what this hook accomplishes, when it should be used, expected param types, etc?

**Review by tay1orjones — CHANGES_REQUESTED** [review]
When `disabled`, I think the icon should probably be modified to match text color?

image

Also how should we treat it with the `light` prop? Selecting between the readonly and light knobs there is no visual difference other than the icon. @carbon-design-system/design 

https://user-images.githubusercontent.com/3360588/120539919-b77b4600-c3ad-11eb-9d48-3cb1f0a041a5.mov

**Comment by janhassel** [comment]
@tay1orjones Actually, now that you bring it up: `readonly` should probably take precedence over `disabled` / block `disabled` from being applied. Or can you think of a use case where a disabled read-only field would make sense?

File: packages/react/src/internal/useNormalizedInputProps.js (hunk 2 above)
Comment (janhassel) [reply packages/react/src/internal/useNormalizedInputProps.js:53]: Pushed an update adding some guidance aorund it. Let me know what you think of it!

**Review by johnbister — COMMENTED** [review]
If read only takes precedence over disabled being applied, it looks fine visually. But if you think of a use case where a disabled read-only would be used, I would make the icon the same color as the disabled text like Taylor said. Let me know what you decide and I’ll approve.

**Comment by johnbister** [comment]
'readonly' taking precedence over 'disabled' looks good for textinput! 

@janhassel I don't see the 'readonly' variant for the passwordinput.  Is this PR still adding the 'readonly' variant to both textinput and passwordinput? I apologize if I'm just missing something.

**Comment by janhassel** [comment]
@johnbister My plan would be to add support for `readonly` step by step to more components. I think building out the ground work in this one and then apply it to others can help keep the PRs manageable.

**Review by johnbister — APPROVED** [review]
lgtm!

**Review by johnbister — APPROVED** [review]
lgtm!

**Comment by janhassel** [comment]
@tay1orjones Regarding the `light` prop question: Afaik the design intent is to not have any background when the input is readonly, so I set it to `transparent`. Therefore there is no visual differenation between `light` and regular.

**Review by tay1orjones — APPROVED** [review]
Thanks for the follow up @janhassel , sounds good 👍

**Pull request merged**